app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -54,33 +97,311 @@ async function waitForPromiseToBeFulfilledAfterRunningAllTimers(
+      });
+    });
+
+    describe('methods that have a param to specify the block', () => {
+      const supportingBlockParam = [
+        { name: 'eth_call', blockParamIndex: 1 },
+        { name: 'eth_getBalance', blockParamIndex: 1 },
+        { name: 'eth_getBlockByNumber', blockParamIndex: 0 },
+        { name: 'eth_getCode', blockParamIndex: 1 },
+        { name: 'eth_getStorageAt', blockParamIndex: 2 },
+        { name: 'eth_getTransactionCount', blockParamIndex: 1 },
+      ];
+      supportingBlockParam.forEach(({ name, blockParamIndex }) => {
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodSupportingBlockParam(name, {
+            providerType,
+            blockParamIndex,
+          });
+        });
+      });
+    });
+
+    describe('methods that assume there is no block param', () => {
+      const assumingNoBlockParam = [
+        'eth_blockNumber',
+        'eth_estimateGas',
+        'eth_gasPrice',
+        'eth_getBlockByHash',
+        // NOTE: eth_getBlockTransactionCountByNumber does take a block param at
+        // the 0th index, but this is not handled by our cache middleware
+        // currently
+        'eth_getBlockTransactionCountByNumber',
+        // NOTE: eth_getTransactionByBlockNumberAndIndex does take a block param
+        // at the 0th index, but this is not handled by our cache middleware
+        // currently
+        'eth_getTransactionByBlockNumberAndIndex',
+        'eth_getBlockTransactionCountByHash',
+        'eth_getFilterLogs',
+        'eth_getTransactionByBlockHashAndIndex',
+        'eth_getUncleByBlockHashAndIndex',
+        // NOTE: eth_getUncleByBlockNumberAndIndex does take a block param at
+        // the 0th index, but this is not handled by our cache middleware
+        // currently
+        'eth_getUncleByBlockNumberAndIndex',
+        'eth_getUncleCountByBlockHash',
+        // NOTE: eth_getUncleCountByBlockNumber does take a block param at the
+        // 0th index, but this is not handled by our cache middleware currently
+        'eth_getUncleCountByBlockNumber',
+      ];
+      assumingNoBlockParam.forEach((name) =>
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodAssumingNoBlockParam(name, { providerType });
+        }),
+      );
+    });
+
+    describe('methods with block hashes in their result', () => {
+      const methodsWithBlockHashInResponse = [
+        'eth_getTransactionByHash',
+        'eth_getTransactionReceipt',
+      ];
+      methodsWithBlockHashInResponse.forEach((method) => {
+        describe(`method name: ${method}`, () => {
+          testsForRpcMethodsThatCheckForBlockHashInResponse(method, {
+            providerType,
+          });
+        });
+      });
+    });
+
+    describe('other methods', () => {
+      describe('eth_getTransactionByHash', () => {
+        it("refreshes the block tracker's current block if it is less than the block number that comes back in the response", async () => {
```

Comment:
@BelfordZ You had placed these in `testsForRpcMethodsThatCheckForBlockHashInResponse`, but I don't think that's the right place for these tests. The purpose of `testsForRpcMethodsThatCheckForBlockHashInResponse` was to test for behavior related to caching (specifically [these lines in the `block-cache` middleware](https://github.com/MetaMask/eth-json-rpc-middleware/blob/254a66cf5f067f35db96c6a5740c8da9a79e495d/src/block-cache.ts#L110-L128) and doesn't have anything to do with `blockHash` being in the `result`. 

This raises a question of organization and whether the way you've chosen is the best approach. Right now you're grouping RPC methods together by which `testsFor*` function is being called. However, this is a bit rigid because it means if you want to add a test for an RPC method you need to repeat the `describe` for that RPC method and possibly place it in some other group. If we instead organize primarily by RPC method, then we are free to place whatever tests we like in the `describe` for that RPC method. So it looks like:

``` javascript
describe('RPC method 1', () => {
  it('test 1', () => {
    // ...
  });

  it('test 2', () => {
    // ...
  });

  it('test 3', () => {
    // ...
  });
})

describe('RPC method 2', () => {
  it('test 1', () => {
    // ...
  });

  it('test 2', () => {
    // ...
  });

  it('test 3', () => {
    // ...
  });
})
```

The `testsFor*` functions are just shorthands for defining a bunch of related tests in one go but at the end of the day you would still be able think of it like the above.

I realize that the reason why you might have to decided to reorganize the tests this way is due to the naming of the `testsFor*` functions. I should come up with better names that explain the behavior being tested by the tests, which they don't do a good job of right now. I will do this in a separate PR, but just wanted to make a note of it in the meantime.

Reply:
We may introduce some changes in a future PR based on work Zach did locally that didn't make it into this PR.